### PR TITLE
updated numpy versions

### PR DESCRIPTION
**Pull request by limlam96**
Loosened numpy pin, was causing compatibility issues

Update - this grew a bit, because of cascading version issues moved to toml approach. To enable work, set up devcontainers config file. Updated github actions to work with new tomls.

Also updated envs introduced need to edit logic to accomodate new shap behaviour

### CODE REVIEW

File: .github/workflows/python-package.yml [hunk 1]
```diff
@@ -16,7 +16,7 @@ jobs:
     runs-on: "ubuntu-22.04"
     strategy:
       matrix:
-        python-version: ["3.8", "3.9"]
+        python-version: ["3.9", "3.10"]
```
Comment (adamsardar) [inline .github/workflows/python-package.yml:19]: Why not pythons up to 3.13?

File: .devcontainer/devcontainer.json [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+	"image": "mcr.microsoft.com/devcontainers/python:0-3.9",
+
+	// Features to add to the dev container. More info: https://containers.dev/features.
+	// "features": {},
+
+	// Use 'forwardPorts' to make a list of ports inside the container available locally.
+	// "forwardPorts": [],
+
+	// Use 'postCreateCommand' to run commands after the container is created.
+	"postCreateCommand": "pip3 install --user -r requirements-dev.txt; pre-commit install",
```
Comment (adamsardar) [inline .devcontainer/devcontainer.json:15]: Will add a ticket, but using uv for a faster startup would improve developer experience (faster loads)

File: pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+]
+
+[project.optional-dependencies]
+dev = [
+    "xgboost",
+    "pytest",
+    "pytest-mock",
+    "pytest-cov",
+    "pytest-benchmark",
+    "bandit",
```
Comment (adamsardar) [inline pyproject.toml:35]: Do you need bandit in here?

File: pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,70 @@
+[build-system]
+requires = [ "setuptools>=61", "wheel", "setuptools-git-versioning>=2.0,<3", ]
+build-backend = "setuptools.build_meta"
+
+[tool.setuptools-git-versioning]
+enabled = true # Allows setuptools-git-versioning to pull version from git state
+
+[tool.setuptools]
+py-modules = []
```
Comment (adamsardar) [inline pyproject.toml:9]: B: Won't this trigger the issue of no exported modules? Rely on autodiscovery instead

File: model_interpreter/interpreter.py [hunk 5]
```diff
@@ -244,22 +247,32 @@ def _get_single_model_contribution(
...
 
-        if isinstance(shap_vals, list):
-            if predict_class_index not in np.arange(
-                len(shap_vals) + 1
-            ) or not isinstance(predict_class_index, int):
-                raise ValueError("predicted class not a valid int")
-
-            shap_values = shap_vals[predict_class_index][0]
-
+        # if values are all 0, shap doesn't bother splitting by class which impacts
```
Comment (adamsardar) [inline model_interpreter/interpreter.py:250]: Not really a fan of the logical flow in this code - very hard to follow. I suspect that it should be handled via generic dispatch on the basis of model type. Will create a follow up ticket

Comment (adamsardar) [reply model_interpreter/interpreter.py:250]: I'm essentially having to rely on the tests. The code seems fine, but the use of magic dimension indicies feels prime for logic errors. But without in depth knowledge of the shap API I find it hard to debug. 

With passing tests and suitable test logic, I'm inclined to accept changes (pending the two small fix items in the review). Clearer code as a follow up ticket.

File: pyproject.toml (hunk 3 above)
Comment (adamsardar) [reply pyproject.toml:35]: Also, consider using xgboost-cpu so that you don't need all the GPU drivers

File: .github/workflows/python-package.yml (hunk 1 above)
Comment (limlam96) [reply .github/workflows/python-package.yml]: added up to 13 here https://github.com/lvgig/model_interpreter/pull/2/commits/43e2f03a3cae4ecd9aeb80018808d654c67ac8a0

but 13 had a mysterious error that i suggest we spinout into another bug/issue, so updated to just go up to 12
https://github.com/lvgig/model_interpreter/pull/2/commits/ad1bc53c9f3afafeaf4cd0b7ef567535d24a6f40

File: pyproject.toml (hunk 3 above)
Comment (limlam96) [reply pyproject.toml]: good catch, general env updates here
https://github.com/lvgig/model_interpreter/pull/2/commits

switched xgboost here https://github.com/lvgig/model_interpreter/pull/2/commits/5ed23c71d88205a44e83368f35c8d75079c9be3f

File: pyproject.toml (hunk 4 above)
Comment (limlam96) [reply pyproject.toml]: I think we do need something as otherwise e.g. the examples folder causes issues, I've updated to copy tubulars setup https://github.com/lvgig/model_interpreter/pull/2/commits

for reference https://github.com/lvgig/tubular/blob/main/pyproject.toml

File: .devcontainer/devcontainer.json (hunk 2 above)
Comment (adamsardar) [reply .devcontainer/devcontainer.json:15]: See #5

File: .github/workflows/python-package.yml (hunk 1 above)
Comment (adamsardar) [reply .github/workflows/python-package.yml]: Logged in #6

File: pyproject.toml (hunk 3 above)
Comment (adamsardar) [reply pyproject.toml]: Nice. Would like to come back to dependencies at some stage - numba is in here, but that drags in all of llvm and is quite beefy. Out of scope though. Thanks for these changes.

File: pyproject.toml [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+build-backend = "setuptools.build_meta"
+
+[tool.setuptools-git-versioning]
+enabled = true # Allows setuptools-git-versioning to pull version from git state
+
+[tool.setuptools]
+packages = ["model_interpreter"]
+
+[project]
+name = "model-interpeter"
```
Comment (adamsardar) [inline pyproject.toml:12]: B: typo - 'model-interpeter' should be 'model-interpreter'

File: pyproject.toml (hunk 4 above)
Comment (adamsardar) [reply pyproject.toml]: I tried to build the package - wowza that sent me off on a goose chase!

`setup.py` makes hard reference to `requirements.txt`, so we were never going to be able to release these changes. So I tweaked what needed to change on the branch package-structure and opened PR #7 with a working config and release pipeline. Either merge it in, or copy the changes that you feel are worthwhile.

I think that your change to py-modules fixed my comment, but we can't release it without fixing the other stuff.

File: pyproject.toml (hunk 6 above)
Comment (limlam96) [reply pyproject.toml]: updated here https://github.com/lvgig/model_interpreter/pull/2/commits

**Review by adamsardar — APPROVED** [review]
